Update extensibility docs (#823)
OpenZeppelin/cairo-contracts: docs/modules/ROOT/pages/usage.adoc

    struct Storage {
        #[substorage(v0)]
        accesscontrol: AccessControlComponent::Storage,
        #[substorage(v0)]
        src5: SRC5Component::Storage
    }

    #[event]
    #[derive(Drop, starknet::Event)]
    enum Event {
        #[flat]
        AccessControlEvent: AccessControlComponent::Event,
        #[flat]
        SRC5Event: SRC5Component::Event
    }

    (...)
}
----

== Customization

:security: xref:/usage.adoc#security[Security]

WARNING: Customizing implementations and accessing component storage can potentially corrupt the state, bypass security checks, and undermine the component logic.
*Exercise extreme caution*. See {security}.

=== Custom implementations

:erc20-component: xref:/api/erc20.adoc#ERC20Component[ERC20Component]
:decimals: xref:/api/erc20.adoc#IERC20Metadata-decimals[decimals]
:ierc20metadata: xref:/api/erc20.adoc#IERC20Metadata[IERC20Metadata]
:custom-decimals: xref:/erc20.adoc#customizing_decimals[Customizing decimals]

There may be instances where a contract requires different behavior from a component implementation.
The contract can create a custom implementation by importing the appropriate interface and defining the new logic.

> In {erc20-component}, for example, the `ERC20MetadataImpl` provides a static {decimals} method which returns`18`.

Reviewer: ```suggestion
In {erc20-component}, for example, the `ERC20MetadataImpl` provides a static {decimals} method which returns `18`.
```

If a contract wanted to change the return value of `decimals`, the `impl` can no longer be embedded.
Rather, the contract would have to create its own `impl` of the {ierc20metadata} interface and expose it in the contract.

> TIP: {custom-decimals} provides a guide for this exact scenario.

Reviewer: i think we need to be way more explicit. this feedback came up twice today (from @ericglau while working in the Wizard, and by some users over telegram).

maybe we should expand this section showing how to make ERC20 transfers pausable

In other words, the `initializable` and `InitializableEvent` names follow this library's convention, but they can be renamed.

=== Storage and events

The component's storage and events must be added to the contract's `Storage` struct and `Event` enum respectively.
If the component doesn't define any events, the compiler will still create an empty event enum inside the component module.

[,javascript]
----
#[starknet::contract]
mod MyContract {
    use openzeppelin::security::InitializableComponent;

    component!(path: InitializableComponent, storage: initializable, event: InitializableEvent);

    #[storage]
    struct Storage {
        #[substorage(v0)]
        initializable: InitializableComponent::Storage
    }

    #[event]
    #[derive(Drop, starknet::Event)]
    enum Event {
        #[flat]
        InitializableEvent: InitializableComponent::Event
    }

    (...)
}
----

The `#[substorage(v0)]` attribute must be included for each component in the `Storage` trait.
This allows the contract to have indirect access to the component's storage.

The `#[flat]` attribute for events in the `Event` enum, however, is not required.

> Component events are not flattened in the component itself because it would remove the event ID from the event log.
> Note that if contracts do not flatten component events, the first key in the event log will be the component ID.
> By flattening the component event in the contract, the first key will be the event ID.

Reviewer: ```suggestion
For component events, the first key in the event log is the component ID.
Flattening the component event removes it, leaving the event ID as the first key.
```
I don't think we need to explain why don't we do it in the component.

Integrating an implementation looks like this:

[,javascript]
----
mod MyContract {
    use openzeppelin::security::InitializableComponent;

    component!(path: InitializableComponent, storage: initializable, event: InitializableEvent);

    (...)

    // Gives the contract access to the implementation methods
    impl InitializableImpl =
        InitializableComponent::InitializableImpl<ContractState>;
}
----

Defining an `impl` gives the contract access to the methods within the implementation from the component.
For example, `is_initialized` is defined in the `InitializableImpl`.
A function on the contract level can expose it like this:

[,javascript]
----
#[starknet::contract]
mod MyContract {
    use openzeppelin::security::InitializableComponent;

    component!(path: InitializableComponent, storage: initializable, event: InitializableEvent);

    (...)

    impl InitializableImpl =
        InitializableComponent::InitializableImpl<ContractState>;

    #[external(v0)]
    fn is_initialized(ref self: ContractState) -> bool {
        self.initializable.is_initialized()
    }
}
----

> Notice that the function must pass the state (`self`) and the component's storage (`initializable`) before finally accessing the implementation method.

Reviewer: I'm not sure what this sentence intends to point out. How to call a method from an impl on the self.storage member? what do you mean by "passing ... the component's storage"?

Author: > I'm not sure what this sentence intends to point out

It's attempting to show the syntax for calling component implementation methods. Maybe there's better names to use when describing it?

> what do you mean by "passing ... the component's storage"?

It's referring to the `storage` arg in the `component!` macro and the member of the `Storage` struct:

```
component!(path: InitializableComponent, storage: initializable, event: InitializableEvent);

#[storage]
struct Storage {
    #[substorage(v0)]
    initializable: InitializableComponent::Storage
}
```

I'll try to rework this and come up with better ways to describe the process

NOTE: Contracts can also (indirectly) access a component's storage. See {accessing-storage}.

While there's nothing wrong with manually exposing methods like in the previous example, this process can be tedious for implementations with many methods.
Fortunately, a contract can embed implementations into the ABI which will expose all of the methods of the implementation.
To embed an implementation, add the `#[abi(embed_v0)]` attribute above the `impl`:

[,javascript]
----
#[starknet::contract]
mod MyContract {
    (...)

    // This attribute exposes the methods of the `impl`
    #[abi(embed_v0)]
    impl InitializableImpl =
        InitializableComponent::InitializableImpl<ContractState>;
}
----

`InitializableImpl` defines the `is_initialized` method in the component.
By adding the embed attribute, `is_initialized` becomes a contract entrypoint for `MyContract`.

> 

Reviewer: since this whole section is about explaining how to use the library, I think it's worth mentioning that this concept guides big part of our API design: we separate implementations between embeddable and internal, making it easy to embed safely. We should mention that and maybe also link to some API doc example -- which can be hard to maintain/not break over time 😅

Author: Good idea. I added it as a TIP at the end of the `Implementations` section. Let me know what you think

=== Initializers

:ownable-component: xref:/api/access.adoc#OwnableComponent[OwnableComponent]

WARNING: Failing to use a component's `initializer` can result in irreparable contract deployments.
Always read the API documentation for each integrated component.

Some components require some sort of setup upon construction.
Usually, this would be a job for a constructor; however, components themselves cannot provide constructors.

> Components instead offer ``initializer``s within their `InternalImpl` which enables a contract to create a constructor and invoke the component's `initializer`.

Reviewer: ```suggestion
Components instead offer ``initializer``s within their `InternalImpl` to be called by the contract's constructor.
```

    component!(path: PausableComponent, storage: pausable, event: PausableEvent);
    component!(path: ERC20Component, storage: erc20, event: ERC20Event);

    #[abi(embed_v0)]
    impl PausableImpl = PausableComponent::PausableImpl<ContractState>;
    impl PausableInternalImpl = PausableComponent::InternalImpl<ContractState>;

    // `ERC20MetadataImpl` can keep the embed directive because the implementation
    // will not change
    #[abi(embed_v0)]
    impl ERC20MetadataImpl = ERC20Component::ERC20MetadataImpl<ContractState>;
    // Do not add the embed directive to these implementations because
    // these will be customized
    impl ERC20Impl = ERC20Component::ERC20Impl<ContractState>;
    impl ERC20CamelOnlyImpl = ERC20Component::ERC20CamelOnlyImpl<ContractState>;

    impl ERC20InternalImpl = ERC20Component::InternalImpl<ContractState>;

    (...)
}
----

The first thing to notice is that the contract imports the interfaces of the implementations that will be customized.
These will be used in the next code example.

Next, the contract includes the {erc20-component} implementations; however, `ERC20Impl` and `ERC20CamelOnlyImplt` are *not* embedded.

> Creating a custom implementation of an interface means that the component implementation cannot be embedded.

Reviewer: ```suggestion
Instead, we want to embed our custom implementation of an interface.
```

    #[external(v0)]
    impl CustomERC20CamelOnlyImpl of IERC20CamelOnly<ContractState> {
        fn totalSupply(self: @ContractState) -> u256 {
            self.erc20.total_supply()
        }

        fn balanceOf(self: @ContractState, account: ContractAddress) -> u256 {
            self.erc20.balance_of(account)
        }

        fn transferFrom(
            ref self: ContractState,
            sender: ContractAddress,
            recipient: ContractAddress,
            amount: u256
        ) -> bool {
            self.pausable.assert_not_paused();
            self.erc20.transfer_from(sender, recipient, amount)
        }
    }
}
----

Notice that in the `CustomERC20Impl`, the `transfer` method integrates `pausable.assert_not_paused` as well as `erc20.transfer` from `PausableImpl` and `ERC20Impl` respectively.
This is why the contract defined the `ERC20Impl` from the component in the previous example.

Creating a custom implementation of an interface must define *all* methods from that interface.
This is true even if the behavior of a method does not change from the component implementation (as `total_supply` exemplifies in this example).

> TIP: The ERC20 documentation provides a more specific custom implementations guide with {custom-decimals}.

Reviewer: ```suggestion
TIP: The ERC20 documentation provides another custom implementation guide for {custom-decimals}.
```
I'm not sure is more specific than pausable ERC20

=== Accessing component storage

Just as contracts can access methods within a component implementation, contracts can also access component storage.

> Storage members are accessible to the contract by instantiating the component's `InternalImpl` like this:

Reviewer: are you sure it's the internalimpl the one that enables this?

Author: Whoops 😅 should be substorage attribute^

[,javascript]
----
#[starknet::contract]
mod MyContract {
    (...)

    impl InternalImpl = InitializableComponent::InternalImpl<ContractState>;

    fn write_to_comp_storage(ref self: ContractState) -> bool {
        self.initializable.Initializable_initialized.write(true);
    }

    fn read_from_comp_storage(self: @ContractState) -> bool {
        self.initializable.Initializable_initialized.read()
    }
}
----

== Security

The maintainers of OpenZeppelin Contracts for Cairo are mainly concerned with the correctness and security of the code as published in the library.

> Customizing implementations and manipulating the component state may break some important assumptions and introduce vulnerabilities in otherwise secure code.

Reviewer: ```suggestion
Customizing implementations and manipulating the component state may break some important assumptions and introduce vulnerabilities.
```
I don't know about that claim

`InitializableImpl` defines the `is_initialized` method in the component.
By adding the embed attribute, `is_initialized` becomes a contract entrypoint for `MyContract`.

[TIP]
====

> Embeddable implementations, when available in this library's components, are segregated from the internal component implementation which makes it easier to safely embed.

Reviewer: ```suggestion
Embeddable implementations, when available in this library's components, are segregated from the internal component implementation which makes it easier to safely expose.
```
alternate with synonyms to not stress word repetition

Some components require some sort of setup upon construction.

> Usually, this would be a job for a constructor; however, components themselves cannot provide constructors.

Reviewer: ```suggestion
Usually, this would be a job for a constructor; however, components themselves cannot implement constructors.
```

> Components instead offer ``initializer``s within their `InternalImpl` to be called by the contract's constructor.

Reviewer: ```suggestion
Components instead offer ``initializer``s within their `InternalImpl` to call from the contract's constructor.
```
i know i suggested the current version but I think these two changes read better

These will be used in the next code example.

Next, the contract includes the {erc20-component} implementations; however, `ERC20Impl` and `ERC20CamelOnlyImplt` are *not* embedded.

> Instead, we want to embed our custom implementation of an interface.

Reviewer: ```suggestion
Instead, we want to expose our custom implementation of an interface.
```
it's not actually embedding (again, my bad)
